### PR TITLE
accept diagnostics as args

### PR DESCRIPTION
**Pull request by deads2k**
Allows specifying diagnostic names as args, but doesn't remove the flag.

@sosiouxme @stevekuznetsov ptal

### CODE REVIEW

File: test/cmd/admin.sh [hunk 1]
```diff
@@ -9,7 +9,7 @@ source "${OS_ROOT}/hack/util.sh"
 source "${OS_ROOT}/hack/cmd_util.sh"
 os::log::install_errexit
 
-# Cleanup cluster resources created by this test
+Cleanup cluster resources created by this test
```
Comment (Miciah) [inline test/cmd/admin.sh]: Where'd the `#` go?

File: test/cmd/admin.sh [hunk 2]
```diff
@@ -299,3 +299,8 @@ os::cmd::expect_success_and_not_text "oc get clusterrolebindings/cluster-admins
 os::cmd::expect_success_and_not_text "oc get rolebindings/cluster-admin         --output-version=v1 --template='{{.subjects}}' -n default" 'cascaded-group'
 os::cmd::expect_success_and_not_text "oc get scc/restricted                     --output-version=v1 --template='{{.groups}}'"              'cascaded-group'
 echo "user-group-cascade: ok"
+
+
+os::cmd::expect_failure_and_text 'openshift ex diagnostics FakeDiagnostic --diagnostics AlsoMissing' 'No requested diagnostics are available: requested=AlsoMissing, FakeDiagnostic'
+os::cmd::expect_failure_and_text 'openshift ex diagnostics AnalyzeLogs --diagnostics AlsoMissing' 'Not all requested diagnostics are available: requested=AlsoMissing, AnalyzeLogs'
+echo "diagnostics: ok"
```
Comment (sosiouxme) [inline test/cmd/admin.sh]: FWIW there is a file for diagnostics test cmds once this merges https://github.com/openshift/origin/pull/6397/files#diff-6061a923ddc8d6dfdd6620834faccb34

**Comment by sosiouxme** [comment]
@deads2k Since there doesn't seem to be a need for subcommands I think this is a good improvement - comma lists suck. In fact if we're going to go this route I'd prefer to just remove `--diagnostics` entirely.

**Comment by deads2k** [comment]
[test]

File: test/cmd/admin.sh (hunk 1 above)
Comment (deads2k) [inline test/cmd/admin.sh]: > Where'd the # go?

fixed

**Comment by deads2k** [comment]
> @deads2k Since there doesn't seem to be a need for subcommands I think this is a good improvement - comma lists suck. In fact if we're going to go this route I'd prefer to just remove --diagnostics entirely.

removed.

**Comment by deads2k** [comment]
Pulled in the deprecation commit too. Any other comments?

**Comment by deads2k** [comment]
https://github.com/openshift/origin/issues/5605 and https://github.com/openshift/origin/issues/6259 re[test]

**Comment by deads2k** [comment]
@stevekuznetsov ptal

File: pkg/cmd/admin/validate/node.go [hunk 3]
```diff
@@ -26,6 +26,8 @@ This command validates that a configuration file intended to be used for a node
 
 	valiateNodeConfigExample = ` // Validate node configuration file
   $ %s openshift.local.config/master/node-config.yaml`
+
+	validateNodeConfigDeprecationMessage = `This command is deprecated and will be removed. Use 'openshift ex diagnostics NodeConfigCheck --node-config=<path/to/config.yaml' instead.`
```
Comment (stevekuznetsov) [inline pkg/cmd/admin/validate/node.go]: erroneous `<` in example path

File: pkg/cmd/experimental/diagnostics/diagnostics.go [hunk 4]
```diff
@@ -116,45 +118,48 @@ func NewCommandDiagnostics(name string, fullName string, out io.Writer) *cobra.C
...
 		return err
 	}
 
-	return nil
-}
+	// If not given master/client config file locations, check if the defaults exist
+	// and adjust the options accordingly:
+	if len(o.MasterConfigLocation) == 0 {
+		if _, err := os.Stat(StandardMasterConfigPath); !os.IsNotExist(err) {
+			o.MasterConfigLocation = StandardMasterConfigPath
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: If someone's got a perfectly fine master config file at a non-default position, forgets to add it as a flag, and has a broken file in the default position, will the diagnostic tool show the path to the file it's checking when it shows errors for that file? We should either do that or announce here that we've found a default and are going to use that.

File: pkg/cmd/experimental/diagnostics/diagnostics.go [hunk 5]
```diff
@@ -116,45 +118,48 @@ func NewCommandDiagnostics(name string, fullName string, out io.Writer) *cobra.C
...
-	warnings := []error{}
-	errors := []error{}
-	diagnostics := []types.Diagnostic{}
+func (o *DiagnosticsOptions) Validate() error {
 	available := availableDiagnostics()
-	if len(o.RequestedDiagnostics) == 0 {
-		o.RequestedDiagnostics = available.List()
-	} else if common := intersection(sets.NewString(o.RequestedDiagnostics...), available); len(common) == 0 {
+
+	if common := intersection(sets.NewString(o.RequestedDiagnostics...), available); len(common) == 0 {
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: we should prefer to use the `Intersection()` provided by `sets.String`

File: pkg/cmd/experimental/diagnostics/diagnostics.go [hunk 6]
```diff
@@ -116,45 +118,48 @@ func NewCommandDiagnostics(name string, fullName string, out io.Writer) *cobra.C
...
 		o.Logger.Error("CED3012", log.EvalTemplate("CED3012", "None of the requested diagnostics are available:\n  {{.requested}}\nPlease try from the following:\n  {{.available}}",
 			log.Hash{"requested": o.RequestedDiagnostics, "available": available.List()}))
-		return false, fmt.Errorf("No requested diagnostics available"), 0, 1
+		return fmt.Errorf("No requested diagnostics are available: requested=%s available=%s", strings.Join(o.RequestedDiagnostics, ", "), strings.Join(available.List(), ", "))
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: long lists look a lot better with the `\nt\t` type format for want/have

File: pkg/cmd/experimental/diagnostics/diagnostics.go [hunk 7]
```diff
@@ -163,21 +168,28 @@ only these are available:
...
-	if len(o.MasterConfigLocation) == 0 {
-		if _, err := os.Stat(StandardMasterConfigPath); !os.IsNotExist(err) {
-			o.MasterConfigLocation = StandardMasterConfigPath
-		}
-	}
-	if len(o.NodeConfigLocation) == 0 {
-		if _, err := os.Stat(StandardNodeConfigPath); !os.IsNotExist(err) {
-			o.NodeConfigLocation = StandardNodeConfigPath
-		}
+		return fmt.Errorf("Not all requested diagnostics are available: requested=%s available=%s", strings.Join(o.RequestedDiagnostics, ", "), strings.Join(available.List(), ", "))
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: much more user friendly to give them the unavailable `Difference()`

File: test/cmd/admin.sh [hunk 8]
```diff
@@ -299,3 +299,8 @@ os::cmd::expect_success_and_not_text "oc get clusterrolebindings/cluster-admins
 os::cmd::expect_success_and_not_text "oc get rolebindings/cluster-admin         --output-version=v1 --template='{{.subjects}}' -n default" 'cascaded-group'
 os::cmd::expect_success_and_not_text "oc get scc/restricted                     --output-version=v1 --template='{{.groups}}'"              'cascaded-group'
 echo "user-group-cascade: ok"
+
+
+os::cmd::expect_failure_and_text 'openshift ex diagnostics FakeDiagnostic AlsoMissing' 'No requested diagnostics are available: requested=FakeDiagnostic. AlsoMissing'
```
Comment (stevekuznetsov) [inline test/cmd/admin.sh]: this should fail on:
```
FakeDiagnostic. AlsoMissing
              ^
```

File: pkg/cmd/experimental/diagnostics/diagnostics.go [hunk 9]
```diff
@@ -163,21 +168,28 @@ only these are available:
...
+func availableDiagnostics() sets.String {
+	available := sets.NewString()
+	available.Insert(availableClientDiagnostics.List()...)
+	available.Insert(availableClusterDiagnostics.List()...)
+	available.Insert(availableHostDiagnostics.List()...)
+	return available
+}
+
+// RunDiagnostics builds diagnostics based on the options and executes them, returning a summary.
+func (o DiagnosticsOptions) RunDiagnostics() (bool, error, int, int) {
```
Comment (stevekuznetsov) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: @sosiouxme not a big fan of `bool, error, int, int` - we have objects capable of expressing "this is a collection of errors, some of which are warnings and some failures". They could potentially be used here to simplify things a bit.

File: pkg/cmd/experimental/diagnostics/diagnostics.go (hunk 4 above)
Comment (stevekuznetsov) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: /cc @sosiouxme

File: test/cmd/admin.sh (hunk 8 above)
Comment (Miciah) [inline test/cmd/admin.sh]: Except parameter 2 to `os::cmd::expect_failure_and_text` is a regex, and `.` matches `,`, so it should work, no? (Cute!)

Comment (stevekuznetsov) [inline test/cmd/admin.sh]: Serendipity... the basis for any good test suite

File: pkg/cmd/experimental/diagnostics/diagnostics.go (hunk 4 above)
Comment (deads2k) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: > If someone's got a perfectly fine master config file at a non-default position, forgets to add it as a flag, and has a broken file in the default position, will the diagnostic tool show the path to the file it's checking when it shows errors for that file? We should either do that or announce here that we've found a default and are going to use that.

Problem for another day.

File: pkg/cmd/experimental/diagnostics/diagnostics.go (hunk 6 above)
Comment (deads2k) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: > long lists look a lot better with the \nt\t type format for want/have

switching to spaces to handle copy/paste correctly.

**Comment by deads2k** [comment]
rebased and comments addressed.

File: pkg/cmd/experimental/diagnostics/diagnostics.go (hunk 4 above)
Comment (sosiouxme) [inline pkg/cmd/experimental/diagnostics/diagnostics.go]: @stevekuznetsov the master file checked is noted in the diagnostic output if anything goes wrong, e.g.:

```
WARN:  [DH0005 from diagnostic MasterConfigCheck@openshift/origin/pkg/diagnostics/host/check_master_config.go:58]
       Validation of master config file '/openshift.local.config/master/master-config.yaml' warned:
       assetConfig.loggingPublicURL: invalid value '', Details: required to view aggregated container logs in the console
       assetConfig.metricsPublicURL: invalid value '', Details: required to view cluster metrics in the console
```

**Comment by stevekuznetsov** [comment]
LGTM after rebase

**Comment by deads2k** [comment]
[merge]

**Pull request merged**